File: examples/dotnet/common/RdpRuntime.cs
// Copyright (c) 2026 solar@heliacal.net
// SPDX-License-Identifier: MIT

using System;
using System.Diagnostics;

namespace RdplibExample.Common
{
    // Owns the process rdplib runtime.
    internal sealed class RdpRuntime : IDisposable
    {
        private IntPtr _runtime;

        internal bool IsOpen
        {
            get { return _runtime != IntPtr.Zero; }
        }

        internal IntPtr NativeHandle
        {
            get { return _runtime; }
        }

        // This is an initial allocator size, not a memory limit.
        internal int Open(uint fastAllocatorBytes = 4u * 1024u * 1024u)
        {
            if (_runtime != IntPtr.Zero)
                return RdplibNative.ErrorBusy;

            return RdplibNative.rdplib_runtime_create(out _runtime, fastAllocatorBytes);
        }

        // Every endpoint must be closed first.
        internal int Close()
        {
            if (_runtime == IntPtr.Zero)
                return RdplibNative.Ok;

            int result = RdplibNative.rdplib_runtime_destroy(_runtime);
            if (result == RdplibNative.Ok)
                _runtime = IntPtr.Zero;

            return result;
        }

        public void Dispose()
        {
            int result = Close();
            Debug.Assert(result == RdplibNative.Ok);
        }
    }
}

File: examples/cpp/common/rdp_runtime.cpp
// Copyright (c) 2026 solar
// SPDX-License-Identifier: MIT

#include "rdp_runtime.h"

#include <cassert>

RDPRuntime::RDPRuntime()
    : m_runtime(nullptr)
{
}

RDPRuntime::~RDPRuntime()
{
    int result = Close();
    assert(result == RDPLIB_OK);
    (void)result;
}

int RDPRuntime::Open(std::uint32_t fast_allocator_bytes)
{
    if (m_runtime != nullptr)
        return RDPLIB_ERROR_BUSY;

    return rdplib_runtime_create(&m_runtime, fast_allocator_bytes);
}

int RDPRuntime::Close()
{
    if (m_runtime == nullptr)
        return RDPLIB_OK;

    int result = rdplib_runtime_destroy(m_runtime);
    if (result == RDPLIB_OK)
        m_runtime = nullptr;

    return result;
}
